src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs

```diff
@@ -57,15 +59,24 @@ public static Checksum Create(Stream stream)
 
     public static Checksum Create<T>(T @object, Action<T, ObjectWriter> writeObject)
     {
-        using var stream = SerializableBytes.CreateWritableStream();
+        // Obtain a writer from the pool
+        var objectWriter = s_objectWriterPool.Allocate();
+        var stream = objectWriter.BaseStream;
 
-        using (var objectWriter = new ObjectWriter(stream, leaveOpen: true))
-        {
-            writeObject(@object, objectWriter);
-        }
+        // Invoke the callback to Write object into objectWriter
+        writeObject(@object, objectWriter);
 
+        // Include validation bytes in the new checksum from the stream
         stream.Position = 0;
-        return Create(stream);
+        var newChecksum = Create(stream);
+
+        // Reset object writer back to it's initial state
+        objectWriter.Reset();
```

Comment:
does `writeValidationBytes` then work properly?  do we basically check if we're at pos0 and wrrite out those bytes (i genuinely don't remember).

Reply:
Changed in commit 2 to just rewrite the validation bytes after the reset instead of having the reset set the position to after the validation bytes. Probably easier to understand that way.